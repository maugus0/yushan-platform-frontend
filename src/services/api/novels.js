import { httpClient } from '../httpClient';
import { processImageUrl } from '../../utils/imageUtils';

// Import fallback image
import fallbackImage from '../../assets/images/novel_default.png';

const BASE_URL = 'https://yushan-backend-staging.up.railway.app/api';
const IMAGE_BASE_URL = 'https://yushan-backend-staging.up.railway.app/images';

// Extract gradient constant for reusability
export const GRADIENT_COLORS = 'linear-gradient(135deg, #6B46C1 0%, #9333EA 50%, #7C3AED 100%)';

export const getNovels = async (params = {}) => {
  try {
    const defaultParams = {
      page: 0,
      size: 10,
      sort: 'createTime',
      order: 'desc',
    };

    const queryParams = { ...defaultParams, ...params };
    const searchParams = new URLSearchParams();

    Object.entries(queryParams).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== '') {
        searchParams.append(key, value);
      }
    });

    const response = await httpClient.get(`${BASE_URL}/novels?${searchParams.toString()}`);
    return response.data;
  } catch (error) {
    console.error('Error fetching novels:', error);
    throw error;
  }
};

// Helper function to handle image URL generation with fallback
const getImageUrl = (coverImgUrl) => {
  return processImageUrl(coverImgUrl, IMAGE_BASE_URL, fallbackImage);
};

// Helper function to transform API data to match component expectations
const transformNovelData = (novels) => {
  return novels.map((novel) => ({
    id: novel.id,
    title: novel.title,
    author: novel.authorUsername,
    cover: getImageUrl(novel.coverImgUrl),
    category: novel.categoryName,
    status: novel.isCompleted ? 'Completed' : 'Ongoing',
    description: novel.synopsis,
    rating: parseFloat(novel.avgRating?.toFixed(1)) || 0,
    chapters: novel.chapterCnt,
    tags: [novel.categoryName],
    // Additional fields from API
    uuid: novel.uuid,
    authorId: novel.authorId,
    categoryId: novel.categoryId,
    wordCnt: novel.wordCnt,
    reviewCnt: novel.reviewCnt,
    viewCnt: novel.viewCnt,
    voteCnt: novel.voteCnt,
    yuanCnt: novel.yuanCnt,
    publishTime: novel.publishTime,
    createTime: novel.createTime,
    updateTime: novel.updateTime,
  }));
};

// Specific functions for different novel categories
export const getWeeklyFeaturedNovels = async () => {
  try {
    const response = await getNovels({
      size: 20,
      sort: 'createTime',
      order: 'desc',
      status: 'PUBLISHED',
    });
    const novels = response.data?.content || [];
    return {
      ...response,
      content: transformNovelData(novels.slice(0, 8)),
    };
  } catch (error) {
    console.error('Error fetching weekly featured novels:', error);
    // Return empty content with proper structure instead of throwing
    return { content: [] };
  }
};

export const getOngoingNovels = async () => {
  try {
    const response = await getNovels({
      size: 20,
      sort: 'createTime',
      order: 'desc',
      status: 'PUBLISHED',
    });

    const allNovels = response.data?.content || [];
    const ongoingNovels = allNovels.filter(
      (novel) => novel.status === 'PUBLISHED' && novel.isCompleted === false
    );

    return {
      ...response,
      content: transformNovelData(ongoingNovels.slice(0, 8)),
    };
  } catch (error) {
    console.error('Error fetching ongoing novels:', error);
    return { content: [] };
  }
};

export const getCompletedNovels = async () => {
  try {
    const response = await getNovels({
      size: 100,
      sort: 'createTime',
      order: 'asc', // Changed to 'asc' to get oldest completed novels as stated in comment
      status: 'PUBLISHED',
    });

    const allNovels = response.data?.content || [];
    const completedNovels = allNovels.filter(
      (novel) => novel.status === 'PUBLISHED' && novel.isCompleted === true
    );

    // Return 8 oldest completed novels (now correctly sorted)
    return {
      content: transformNovelData(completedNovels.slice(0, 8)),
    };
  } catch (error) {
    console.error('Error fetching completed novels:', error);
    return { content: [] };
  }
};

export const getNewestNovels = async () => {
  try {
    const response = await getNovels({
      size: 3,
      sort: 'createTime',
      order: 'desc',
      status: 'PUBLISHED',
    });
    return {
      ...response,
      content: transformNovelData(response.data?.content || []),
    };
  } catch (error) {
    console.error('Error fetching newest novels:', error);
    return { content: [] };
  }
};
