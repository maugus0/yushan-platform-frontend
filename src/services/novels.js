import { http, authHeader, toAbsoluteUrl } from './_http';

const novelsApi = {
  async getDetail(id) {
    const res = await http.get(`/novels/${id}`, { headers: authHeader() });
    return res?.data?.data;
  },
  async vote(novelId) {
    const res = await http.post(`/novels/${novelId}/vote`, {}, { headers: authHeader() });
    return res?.data?.data; // { novelId, voteCount, remainedYuan }
  },
  // utility for cover url
  coverUrl(v) {
    return toAbsoluteUrl(v);
  },
  async getChapters(novelId) {
    const res = await http.get(`/chapters/novel/${novelId}?page=1&pageSize=1&publishedOnly=true`, {
      headers: authHeader(),
    });
    return res?.data?.data;
  },
  async getChaptersFull(novelId) {
    const res = await http.get(
      `/chapters/novel/${novelId}?page=1&pageSize=1000&publishedOnly=true`,
      { headers: authHeader() }
    );
    return res?.data?.data;
  },
  async getChapterContent(novelId, chapterNumber) {
    const res = await http.get(`/chapters/novel/${novelId}/number/${chapterNumber}`, {
      headers: authHeader(),
    });
    return res?.data?.data;
  },
  async getChapterByUuid(uuid) {
    const res = await http.get(`/chapters/${uuid}`, { headers: authHeader() });
    return res?.data?.data;
  },
};

export default novelsApi;
