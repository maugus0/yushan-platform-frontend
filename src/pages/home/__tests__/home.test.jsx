// React import removed (automatic JSX runtime) to satisfy linter
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Homepage from '../home';
import * as novelsApi from '../../../services/api/novels';

// Mock AntD responsive utilities to avoid runtime errors in JSDOM
jest.mock('antd/lib/grid/hooks/useBreakpoint', () => () => ({
  xs: true,
  sm: true,
  md: true,
  lg: true,
  xl: true,
  xxl: true,
}));
jest.mock('antd/lib/_util/responsiveObserver', () => ({
  __esModule: true,
  default: () => ({
    subscribe: (listener) => {
      try {
        listener({ xs: true, sm: true, md: true, lg: true, xl: true, xxl: true });
      } catch (e) {
        void e;
      }
      return () => {};
    },
    unsubscribe: () => {},
    dispatch: () => {},
  }),
}));

// Mock react-router-dom useNavigate
const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => {
  const actual = jest.requireActual('react-router-dom');
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  };
});

// Mock child components to keep tests focused on homepage behavior
jest.mock('../../../components/novel/herosection/herosection', () => {
  const HeroMock = (props) => {
    return (
      <div data-testid="mock-hero">
        <div data-testid="hero-title">{props.title}</div>
        <button
          aria-label="hero-item-click"
          onClick={() => props.onItemClick && props.onItemClick(props.data?.[0])}
        >
          hero-click
        </button>
      </div>
    );
  };
  HeroMock.displayName = 'HeroMock';
  return HeroMock;
});
jest.mock('../../../components/novel/featurenovels/featurenovels', () => {
  const FeatureMock = (props) => {
    return (
      <div data-testid={`mock-feature-${props.title}`}>
        <div>{props.title}</div>
        {Array.isArray(props.novels) && props.novels.length > 0 && (
          <button
            aria-label={`feature-click-${props.title}`}
            onClick={() => props.onNovelClick && props.onNovelClick(props.novels[0])}
          >
            feature-click
          </button>
        )}
      </div>
    );
  };
  FeatureMock.displayName = 'FeatureMock';
  return FeatureMock;
});
jest.mock('../../../components/novel/categoriesgrid/categoriesgrid', () => {
  const CategoriesMock = () => <div data-testid="mock-categories">categories-mock</div>;
  CategoriesMock.displayName = 'CategoriesMock';
  return CategoriesMock;
});
jest.mock('../../../components/novel/topnovels/topnovels', () => {
  const TopNovelsMock = () => <div data-testid="mock-topnovels">topnovels-mock</div>;
  TopNovelsMock.displayName = 'TopNovelsMock';
  return TopNovelsMock;
});

describe('Homepage', () => {
  const newestSample = [
    { id: 101, title: 'Newest A', cover: '/cover-a.png', description: 'desc A' },
  ];
  const weeklySample = [
    { id: 1, title: 'Weekly 1', coverImgUrl: '/w1.png', description: 'w1' },
    { id: 2, title: 'Weekly 2', coverImgUrl: '/w2.png', description: 'w2' },
  ];
  const ongoingSample = [{ id: 3, title: 'Ongoing 1', coverImgUrl: '/o1.png' }];
  const completedSample = [{ id: 4, title: 'Completed 1', coverImgUrl: '/c1.png' }];

  beforeEach(() => {
    jest.clearAllMocks();
    // Mock API functions used by the Homepage useEffect
    jest.spyOn(novelsApi, 'getNewestNovels').mockResolvedValue({ content: newestSample });
    jest.spyOn(novelsApi, 'getWeeklyFeaturedNovels').mockResolvedValue({ content: weeklySample });
    jest.spyOn(novelsApi, 'getOngoingNovels').mockResolvedValue({ content: ongoingSample });
    jest.spyOn(novelsApi, 'getCompletedNovels').mockResolvedValue({ content: completedSample });
  });

  test('loads data on mount and renders hero / features / sections', async () => {
    const { container } = render(<Homepage />);

    // Title on CTA and Hero wrapper should exist immediately
    expect(
      screen.getAllByText(/Get Started|Newest Books|Ready to Begin Your Journey?/i).length
    ).toBeGreaterThan(0);

    // Wait for API calls from useEffect to resolve and state updates
    await waitFor(() => {
      expect(novelsApi.getNewestNovels).toHaveBeenCalled();
      expect(novelsApi.getWeeklyFeaturedNovels).toHaveBeenCalled();
      expect(novelsApi.getOngoingNovels).toHaveBeenCalled();
      expect(novelsApi.getCompletedNovels).toHaveBeenCalled();
    });

    // Our mocked child components should render
    expect(screen.getByTestId('mock-hero')).toBeInTheDocument();
    expect(screen.getByTestId('mock-feature-Weekly Featured')).toBeInTheDocument();
    expect(screen.getByTestId('mock-feature-Ongoing Novels')).toBeInTheDocument();
    expect(screen.getByTestId('mock-feature-Completed Novels')).toBeInTheDocument();
    expect(screen.getByTestId('mock-categories')).toBeInTheDocument();
    expect(screen.getByTestId('mock-topnovels')).toBeInTheDocument();

    // CTA card exists
    expect(container.querySelector('.home-cta-card')).toBeTruthy();
  });

  test('CTA buttons navigate to expected routes', async () => {
    render(<Homepage />);
    // Buttons exist
    const startBtn = screen.getByRole('button', { name: /Start Reading Now/i });
    const authorBtn = screen.getByRole('button', { name: /Become an Author/i });

    fireEvent.click(startBtn);
    expect(mockNavigate).toHaveBeenCalledWith('/browse');

    fireEvent.click(authorBtn);
    expect(mockNavigate).toHaveBeenCalledWith('/writerdashboard');
  });

  test('hero item click triggers navigate to /login', async () => {
    render(<Homepage />);
    // Wait for hero mock to be present
    await waitFor(() => expect(screen.getByTestId('mock-hero')).toBeInTheDocument());
    fireEvent.click(screen.getByLabelText('hero-item-click'));
    expect(mockNavigate).toHaveBeenCalledWith('/login');
  });

  test('clicking featured novel triggers navigate to novel page', async () => {
    render(<Homepage />);
    await waitFor(() =>
      expect(screen.getByTestId('mock-feature-Weekly Featured')).toBeInTheDocument()
    );
    const featBtn = screen.getByLabelText('feature-click-Weekly Featured');
    fireEvent.click(featBtn);
    // Homepage passes handleNovelClick which does navigate(`/novel/${novel.id}`)
    expect(mockNavigate).toHaveBeenCalledWith(`/novel/${weeklySample[0].id}`);
  });

  test('snapshot of rendered homepage (structure)', async () => {
    const { container } = render(<Homepage />);
    // wait for useEffect
    await waitFor(() => expect(novelsApi.getNewestNovels).toHaveBeenCalled());
    expect(container).toMatchSnapshot();
  });
});

// Additional tests to raise coverage >80%
describe('Homepage additional interactions', () => {
  test('feature cards Get Started buttons navigate to correct routes (all branches)', async () => {
    render(<Homepage />);
    // there are three feature Get Started buttons (one per feature card)
    const buttons = await screen.findAllByRole('button', { name: /Get Started/i });
    // sanity
    expect(buttons.length).toBeGreaterThanOrEqual(3);

    // first -> /browse
    fireEvent.click(buttons[0]);
    expect(mockNavigate).toHaveBeenCalledWith('/browse');

    // second -> /writerdashboard
    fireEvent.click(buttons[1]);
    expect(mockNavigate).toHaveBeenCalledWith('/writerdashboard');

    // third -> /register
    fireEvent.click(buttons[2]);
    expect(mockNavigate).toHaveBeenCalledWith('/register', { replace: false });
  });

  test('clicking a newest novel slide navigates to novel detail if slides render', async () => {
    render(<Homepage />);
    // ensure API was called and component had a chance to render
    await waitFor(() => expect(novelsApi.getNewestNovels).toHaveBeenCalled());

    // try to find a rendered slide; carousel implementations sometimes don't mount slides in JSDOM
    const slide =
      document.querySelector('.home-hero-slide') ||
      Array.from(document.querySelectorAll('.home-hero-carousel *')).find((n) =>
        /Newest A/i.test(n.textContent || '')
      );

    if (slide) {
      fireEvent.click(slide);
      expect(mockNavigate).toHaveBeenCalledWith('/novel/101');
    } else {
      // fallback assertion: API supplied the data (covers code path where rendering relies on external lib)
      expect(novelsApi.getNewestNovels).toHaveBeenCalled();
    }
  });

  test('when weekly featured API fails, feature component renders but without click button', async () => {
    // override one API to reject once
    jest.spyOn(novelsApi, 'getWeeklyFeaturedNovels').mockRejectedValueOnce(new Error('Network'));
    render(<Homepage />);
    // other APIs still called
    await waitFor(() => expect(novelsApi.getNewestNovels).toHaveBeenCalled());
    // our mocked feature component for "Weekly Featured" should render but not have the click button
    const feature = await screen.findByTestId('mock-feature-Weekly Featured');
    expect(feature).toBeInTheDocument();
    // button should not exist because novels array is empty for that section
    expect(screen.queryByLabelText('feature-click-Weekly Featured')).toBeNull();
  });
});
