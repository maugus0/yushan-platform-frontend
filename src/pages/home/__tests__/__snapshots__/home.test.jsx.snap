// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homepage snapshot of rendered homepage (structure) 1`] = `
<div>
  <div
    class="home-bg"
  >
    <section
      class="home-hero-section"
    >
      <div
        class="home-hero-flex"
      >
        <div
          class="home-hero-flex-left"
        >
          <div
            class="home-hero-carousel-wrapper"
          >
            <div
              class="home-hero-carousel-title"
            >
              <h3
                class="ant-typography home-hero-carousel-title-text css-dev-only-do-not-override-1odpy5d"
              >
                Newest Books
              </h3>
            </div>
            <div
              class="ant-carousel css-dev-only-do-not-override-1odpy5d"
            >
              <div
                class="slick-slider home-hero-carousel slick-initialized"
                dir="ltr"
              >
                <div
                  class="slick-list"
                >
                  <div
                    class="slick-track"
                    style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
                  >
                    <div
                      aria-hidden="false"
                      class="slick-slide slick-active slick-current"
                      data-index="0"
                      style="outline: none; width: 0px;"
                      tabindex="-1"
                    >
                      <div>
                        <div
                          style="width: 100%; display: inline-block;"
                          tabindex="-1"
                        >
                          <div
                            class="home-hero-slide"
                            style="position: relative; justify-content: flex-start; cursor: pointer;"
                          >
                            <div
                              class="home-hero-blur-bg"
                              style="opacity: 0.9;"
                            />
                            <div
                              class="home-hero-slide-content home-hero-slide-content-left"
                            >
                              <div
                                class="home-hero-img"
                              >
                                <img
                                  alt="Newest A"
                                  class="home-hero-img-el"
                                  src="/cover-a.png"
                                />
                              </div>
                              <div
                                class="home-hero-content"
                              >
                                <h2
                                  class="ant-typography home-hero-title home-hero-title-white css-dev-only-do-not-override-1odpy5d"
                                >
                                  Newest A
                                </h2>
                                <div
                                  class="ant-typography home-hero-desc home-hero-desc-white css-dev-only-do-not-override-1odpy5d"
                                >
                                  desc A
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <ul
                  class="slick-dots slick-dots-bottom"
                  style="display: block;"
                >
                  <li
                    class="slick-active"
                  >
                    <button>
                      1
                    </button>
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </div>
        <div
          class="home-hero-flex-right"
        >
          <div
            data-testid="mock-hero"
          >
            <div
              data-testid="hero-title"
            >
              About Yushan
            </div>
            <button
              aria-label="hero-item-click"
            >
              hero-click
            </button>
          </div>
        </div>
      </div>
    </section>
    <section
      class="home-features-section"
    >
      <div
        class="home-features-flex"
      >
        <div
          class="home-feature-card-wrapper"
        >
          <div
            class="ant-card ant-card-bordered ant-card-hoverable home-feature-card css-dev-only-do-not-override-1odpy5d"
            style="padding: 32px 24px; background-color: rgb(240, 248, 255);"
          >
            <div
              class="ant-card-body"
            >
              <div
                class="ant-space css-dev-only-do-not-override-1odpy5d ant-space-vertical ant-space-gap-row-large ant-space-gap-col-large"
                style="width: 100%;"
              >
                <div
                  class="ant-space-item"
                >
                  <div>
                    <span
                      aria-label="book"
                      class="anticon anticon-book"
                      role="img"
                      style="font-size: 48px; color: rgb(24, 144, 255);"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="book"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M832 64H192c-17.7 0-32 14.3-32 32v832c0 17.7 14.3 32 32 32h640c17.7 0 32-14.3 32-32V96c0-17.7-14.3-32-32-32zm-260 72h96v209.9L621.5 312 572 347.4V136zm220 752H232V136h280v296.9c0 3.3 1 6.6 3 9.3a15.9 15.9 0 0022.3 3.7l83.8-59.9 81.4 59.4c2.7 2 6 3.1 9.4 3.1 8.8 0 16-7.2 16-16V136h64v752z"
                        />
                      </svg>
                    </span>
                  </div>
                </div>
                <div
                  class="ant-space-item"
                >
                  <h3
                    class="ant-typography home-feature-title css-dev-only-do-not-override-1odpy5d"
                  >
                    Read Novels
                  </h3>
                </div>
                <div
                  class="ant-space-item"
                >
                  <div
                    class="ant-typography home-feature-desc css-dev-only-do-not-override-1odpy5d"
                  >
                    Discover thousands of captivating web novels across all genres. From cultivation to romance, fantasy to sci-fi.
                  </div>
                </div>
                <div
                  class="ant-space-item"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-1odpy5d ant-btn-primary ant-btn-color-primary ant-btn-variant-solid ant-btn-lg home-feature-btn"
                    type="button"
                  >
                    <span>
                      Get Started
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="home-feature-card-wrapper"
        >
          <div
            class="ant-card ant-card-bordered ant-card-hoverable home-feature-card css-dev-only-do-not-override-1odpy5d"
            style="padding: 32px 24px; background-color: rgb(246, 255, 237);"
          >
            <div
              class="ant-card-body"
            >
              <div
                class="ant-space css-dev-only-do-not-override-1odpy5d ant-space-vertical ant-space-gap-row-large ant-space-gap-col-large"
                style="width: 100%;"
              >
                <div
                  class="ant-space-item"
                >
                  <div>
                    <span
                      aria-label="edit"
                      class="anticon anticon-edit"
                      role="img"
                      style="font-size: 48px; color: rgb(82, 196, 26);"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="edit"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 000-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 009.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"
                        />
                      </svg>
                    </span>
                  </div>
                </div>
                <div
                  class="ant-space-item"
                >
                  <h3
                    class="ant-typography home-feature-title css-dev-only-do-not-override-1odpy5d"
                  >
                    Write Novels
                  </h3>
                </div>
                <div
                  class="ant-space-item"
                >
                  <div
                    class="ant-typography home-feature-desc css-dev-only-do-not-override-1odpy5d"
                  >
                    Share your stories with the world. Our platform provides all the tools you need to publish and promote your work.
                  </div>
                </div>
                <div
                  class="ant-space-item"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-1odpy5d ant-btn-primary ant-btn-color-primary ant-btn-variant-solid ant-btn-lg home-feature-btn"
                    type="button"
                  >
                    <span>
                      Get Started
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="home-feature-card-wrapper"
        >
          <div
            class="ant-card ant-card-bordered ant-card-hoverable home-feature-card css-dev-only-do-not-override-1odpy5d"
            style="padding: 32px 24px; background-color: rgb(255, 251, 230);"
          >
            <div
              class="ant-card-body"
            >
              <div
                class="ant-space css-dev-only-do-not-override-1odpy5d ant-space-vertical ant-space-gap-row-large ant-space-gap-col-large"
                style="width: 100%;"
              >
                <div
                  class="ant-space-item"
                >
                  <div>
                    <span
                      aria-label="trophy"
                      class="anticon anticon-trophy"
                      role="img"
                      style="font-size: 48px; color: rgb(250, 173, 20);"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="trophy"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M868 160h-92v-40c0-4.4-3.6-8-8-8H256c-4.4 0-8 3.6-8 8v40h-92a44 44 0 00-44 44v148c0 81.7 60 149.6 138.2 162C265.7 630.2 359 721.7 476 734.5v105.2H280c-17.7 0-32 14.3-32 32V904c0 4.4 3.6 8 8 8h512c4.4 0 8-3.6 8-8v-32.3c0-17.7-14.3-32-32-32H548V734.5C665 721.7 758.3 630.2 773.8 514 852 501.6 912 433.7 912 352V204a44 44 0 00-44-44zM184 352V232h64v207.6a91.99 91.99 0 01-64-87.6zm520 128c0 49.1-19.1 95.4-53.9 130.1-34.8 34.8-81 53.9-130.1 53.9h-16c-49.1 0-95.4-19.1-130.1-53.9-34.8-34.8-53.9-81-53.9-130.1V184h384v296zm136-128c0 41-26.9 75.8-64 87.6V232h64v120z"
                        />
                      </svg>
                    </span>
                  </div>
                </div>
                <div
                  class="ant-space-item"
                >
                  <h3
                    class="ant-typography home-feature-title css-dev-only-do-not-override-1odpy5d"
                  >
                    Earn Yuan/XP and Level Up
                  </h3>
                </div>
                <div
                  class="ant-space-item"
                >
                  <div
                    class="ant-typography home-feature-desc css-dev-only-do-not-override-1odpy5d"
                  >
                    Earn rewards for reading, writing, and engaging with the community. Level up your profile and unlock exclusive features.
                  </div>
                </div>
                <div
                  class="ant-space-item"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-1odpy5d ant-btn-primary ant-btn-color-primary ant-btn-variant-solid ant-btn-lg home-feature-btn"
                    type="button"
                  >
                    <span>
                      Get Started
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
    <section
      class="home-cta-section"
      style="padding: 80px 0px;"
    >
      <div
        class="ant-row ant-row-center css-dev-only-do-not-override-1odpy5d"
      >
        <div
          class="ant-col ant-col-xs-24 ant-col-sm-20 ant-col-md-16 ant-col-lg-12 css-dev-only-do-not-override-1odpy5d"
          style="text-align: center;"
        >
          <div
            class="ant-card ant-card-bordered home-cta-card css-dev-only-do-not-override-1odpy5d"
            style="background-color: rgba(255, 255, 255, 0.1);"
          >
            <div
              class="ant-card-body"
            >
              <h2
                class="ant-typography home-cta-title css-dev-only-do-not-override-1odpy5d"
                style="color: rgb(26, 26, 26);"
              >
                Ready to Begin Your Journey?
              </h2>
              <div
                class="ant-typography home-cta-desc css-dev-only-do-not-override-1odpy5d"
                style="color: rgb(74, 85, 104);"
              >
                Join thousands of readers and writers in the Yushan community. Your next favorite story is just a click away.
              </div>
              <div
                class="ant-space css-dev-only-do-not-override-1odpy5d ant-space-horizontal ant-space-align-center ant-space-gap-row-middle ant-space-gap-col-middle"
                style="flex-wrap: wrap;"
              >
                <div
                  class="ant-space-item"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-1odpy5d ant-btn-primary ant-btn-color-primary ant-btn-variant-solid ant-btn-lg home-cta-btn"
                    style="background-color: rgb(102, 126, 234); border-color: #667eea; color: white;"
                    type="button"
                  >
                    <span>
                      Start Reading Now
                    </span>
                  </button>
                </div>
                <div
                  class="ant-space-item"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-1odpy5d ant-btn-default ant-btn-color-default ant-btn-variant-outlined ant-btn-lg home-cta-link"
                    style="background-color: transparent; border-color: #667eea; color: rgb(102, 126, 234);"
                    type="button"
                  >
                    <span>
                      Become an Author
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
    <div
      data-testid="mock-feature-Weekly Featured"
    >
      <div>
        Weekly Featured
      </div>
      <button
        aria-label="feature-click-Weekly Featured"
      >
        feature-click
      </button>
    </div>
    <div
      data-testid="mock-categories"
    >
      categories-mock
    </div>
    <div
      data-testid="mock-feature-Ongoing Novels"
    >
      <div>
        Ongoing Novels
      </div>
      <button
        aria-label="feature-click-Ongoing Novels"
      >
        feature-click
      </button>
    </div>
    <div
      data-testid="mock-topnovels"
    >
      topnovels-mock
    </div>
    <div
      data-testid="mock-feature-Completed Novels"
    >
      <div>
        Completed Novels
      </div>
      <button
        aria-label="feature-click-Completed Novels"
      >
        feature-click
      </button>
    </div>
  </div>
</div>
`;
